qiskit/pulse/compiler/target.py

```diff
@@ -0,0 +1,306 @@
+                of the qubit control signal.
+            meas_frames: A dictionary of measurement frame identifier keyed on
+                the Qiskit :class:`.MeasurementFrame` object index.
+                This frame must track the rotating frame
+                of the measurement stimulus signal.
+            qubit_ports: A dictioanry of hardware port identifier keyed on
+                the Qiskit :class:`~qiskit.pulse.model.Qubit` object index.
+                This port must be used to drive qubit regardless of frames.
+            coupler_ports: A dictionary of hardware port identifier keyed on
+                the Qiskit :class:`.Port` object index.
+                This port must be used to drive multi-qubit interactions.
+            mixed_frams: A dictionary of avilable mixed frame resources keyed on
+                the unique identifier of the port. Values are list of frame identifiers
+                available for this port to form a mixed frame.
+
+        """
+        self._qubit_frames = qubit_frames
+        self._qubit_frames_inv = dict(zip(qubit_frames.values(), qubit_frames.keys()))
+
+        self._meas_frames = meas_frames
+        self._meas_frames_inv = dict(zip(meas_frames.values(), meas_frames.keys()))
+
+        self._qubit_ports = qubit_ports
+        self._qubit_ports_inv = dict(zip(qubit_ports.values(), qubit_ports.keys()))
+
+        self._coupler_ports = coupler_ports
+        self._coupler_ports_inv = dict(zip(coupler_ports.values(), coupler_ports.keys()))
+
+        self._mixed_frames = mixed_frames
+
+    def is_mixed_frame_available(
+        self,
+        mixed_frame: model.MixedFrame,
+    ) -> bool:
+        if not isinstance(mixed_frame, model.MixedFrame):
+            raise TypeError(f"{mixed_frame} is not a MixedFrame object.")
+        try:
+            p_uid = self.get_port_identifier(mixed_frame.pulse_target)
+            f_uid = self.get_frame_identifier(mixed_frame.frame)
+        except NotExistingComponent:
+            return False
+        try:
+            return f_uid in self._mixed_frames[p_uid]
+        except KeyError:
+            return False
+
+    def get_frame_identifier(
+        self,
+        frame: model.Frame,
+    ) -> str:
+        if isinstance(frame, model.QubitFrame):
+            try:
+                return self._qubit_frames[frame.index]
+            except KeyError as ex:
+                raise NotExistingComponent(
+                    "This hardware doesn't provide any frame for "
+                    f"QubitFrame of index {frame.index}."
+                ) from ex
+        if isinstance(frame, model.MeasurementFrame):
+            try:
+                return self._meas_frames[frame.index]
+            except KeyError as ex:
+                raise NotExistingComponent(
+                    "This hardware doesn't provide any frame for "
+                    f"MeasurementFrame of index {frame.index}."
+                ) from ex
+        raise TypeError(
```

Comment:
I see how the framework here allows for the pulse compiler to map `Qubit` and `Coupler` logical elements and `QubitFrame` and `MeasurementFrame` frames to hardware port and frame identifiers. What about defining additional logical elements or frames? This framewrok lists the port and frame identifiers, but how should the user set up the mapping? Does the user use the indentifiers from `PulseTarget` to create `Port` and `GenericFrame` instances with those identifiers as names to create `MixedFrame`s and then the compiler will translate those directly without look up with the `PulseTarget` methods? Or should this method also have a clause for `GenericFrame` that returns `frame.name`, and should `get_port_identifier` similarly return `port.name` for a `Port` input?

Reply:
I updated the model in [214185d](https://github.com/Qiskit/qiskit/pull/12096/commits/214185d6a5cb0df7cc3fb5e9aeed9fe7ef7ef916). `GenericFrame` and `Port` are also considered. Since end-user cannot define ports, there is validation for port UID inside target. On the other hand, there is no UID validation for frame since I assume they can use arbitrary name. Of course there is limitation in number of available hardware resources, but this is something compiler pass must validate.